docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+    new item to the front of the existing queue, and replaces the head pointer
+    with the new item.
+
+  - **To scan the queue**, a reader acquires the read lock, captures the head
+    pointer, and then releases the lock. The rest of its request can be served
+    without holding a lock, since the queue structure will not change.
+
+	When a reader wants to wait, it will yield the lock and wait on a condition
+	that is signaled when the publisher swings the pointer.
+
+  - **To prune the queue**, the publisher (who is the sole writer) will track
+    the queue length and the age of the oldest item separately.  When the
+    length and or age exceed the configured bounds, it will construct a new
+    queue spine on the same items, discarding out-of-band values.
+
+    Pruning can be done while the publisher already holds the write lock, or
+    could be done outside the lock entirely: Once the new queue is constructed,
+    the lock can be re-acquired to swing the pointer. This costs some extra
+    allocations for the cons cells, but avoids duplicating any event items.
+    The pruning step is a simple linear scan down the first (up to) max-items
+    elements of the queue, to find the breakpoint of age and length.
+
+    Moreover, the publisher can amortize the cost of pruning by item count, if
+    necessary, by pruning length "more aggressively" than the configuration
+    requires (e.g., reducing to 3/4 of the maximum rather than 1/1).
+
+    The state of the event log before the publisher acquires the lock:
+    ![Before publish and pruning](./img/adr-075-log-before.png)
+
+	After the publisher has added a new item and pruned old ones:
+    ![After publish and pruning](./img/adr-075-log-after.png)
+
+### Migration Plan
+
+This design requires that clients eventually migrate to the new event
+subscription API, but provides a full release cycle with both APIs in place to
+make this burden more tractable. The migration strategy is broadly:
+
+**Phase 1**: Release v0.36.
+
+- Implement the new `events` endpoint, keeping the existing methods as they are.
+- Update the Go clients to support the new `events` endpoint, and handle polling.
+- Update the old endpoints to log annoyingly about their own deprecation.
+- Write tutorials about how to migrate client usage.
+
+At or shortly after release, we should proactively update the Cosmos SDK to use
+the new API, to remove a disincentive to upgrading.
+
+**Phase 2**: Release v0.37
+
+- During development, we should actively seek out any existing users of the
+  streaming event subscription API and help them migrate.
+- Possibly also: Spend some time writing clients for JS, Rust, et al.
+- Release: Delete the old implementation and all the websocket support code.
+
+> **Open question**: Even though the plan is to keep the existing service, we
+> might take the opportunity to restrict the websocket endpoint to _only_ the
+> event streaming service, removing the other endpoints.
```

Comment:
This makes sense to me. I'm unsure if there are actually any clients out there using the websocket for endpoints that aren't the event streaming service (and if so why?)

Reply:
I'm not sure either. 

As for "why": There's no strong technical reason to _prefer_ websocket for other requests, but if you already have a websocket client connected, you might use that instead of creating a separate client.

I don't know if anyone actually does this. In Go it would be a relatively easy fix if they did, since the API for other RPC methods isn't changing—they'd just have to create one more client object and update the call site to use that one instead of the websocket client. For non-Go clients, it'd look different but would be about the same effort.

Reply:
I updated this point in light of the above. I think if we want to fit this into v0.36, we should skip this step. If we wind up punting to v0.37, we can re-evaluate.